### PR TITLE
add resize and watermark

### PR DESCRIPTION
**Pull request by Zstro**

### CODE REVIEW

File: imagick/process.go [hunk 1]
```diff
@@ -1,10 +1,60 @@
...
+
+type WatermarkPlan struct {
+	Transparency int
+	Position     string
+	XMargin      int
+	YMargin      int
+	Voffset      int
+	PictureMask  WatermarkPicture
+	TextMask     WatermarkText
+	Order        int //default 0 图片水印在前 1 文字水印在前
```
Comment (fireworkmarks) [inline imagick/process.go]: Change to English comment

File: imagick/process.go [hunk 2]
```diff
@@ -18,6 +68,9 @@ func Initialize() Library {
 }
 
 func (img *ImageWand) Terminate() {
+	img.MagickWand.Destroy()
```
Comment (fireworkmarks) [inline imagick/process.go:71]: Rewrite the instance method Destroy or DestoryImageWand

File: imagick/process.go [hunk 3]
```diff
@@ -29,3 +82,239 @@ func NewImageWand() (lib Library) {
...
+		o.Zoom = factor
+		err = img.resize(o)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	switch plan.Mode {
+	//长边优先
```
Comment (fireworkmarks) [inline imagick/process.go]: English

File: imagick/process.go [hunk 4]
```diff
@@ -29,3 +82,239 @@ func NewImageWand() (lib Library) {
...
+		if err != nil {
+			helper.Log.Error("open watermark picture file failed")
+			return err
+		}
+		wmWidth := int(picture.GetImageWidth())
+		wmHeight := int(picture.GetImageHeight())
+
+		w.Picture = picture
+		w.Transparency = plan.Transparency
+		switch plan.Position {
```
Comment (fireworkmarks) [inline imagick/process.go]: Encapsulatable method

File: imagick/resize.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
+package imagick
+
+import (
+	"github.com/journeymidnight/pipa/helper"
+	"github.com/journeymidnight/pipa/library"
+	"gopkg.in/gographics/imagick.v3/imagick"
+	"math"
+)
+
+const Method = imagick.FILTER_LANCZOS
```
Comment (fireworkmarks) [inline imagick/resize.go:9]: If it is zoomed in, please use "MITCHELL" algorithm, see related analysis for details

File: imagick/resize.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+	if o.Limit && !o.Force {
+		if originWidth < o.Width && originHeight < o.Height {
+			factor = 1.0
+			o.Width = originWidth
+			o.Height = originHeight
+		}
+	}
+
+	if o.Zoom != library.Zoom {
```
Comment (fireworkmarks) [inline imagick/resize.go]: I suggest using switch after abstracting the method name, which is logically a choice

File: imagick/process.go [hunk 7]
```diff
@@ -29,3 +82,239 @@ func NewImageWand() (lib Library) {
...
+			return err
+		}
+		return nil
+	}
+
+	switch plan.Mode {
+	//长边优先
+	case "lfit":
+		adjustCropTask(plan, img.MagickWand.GetImageWidth(), img.MagickWand.GetImageHeight())
+		o.Width = plan.Width
```
Comment (fireworkmarks) [inline imagick/process.go:137]: repeat

File: imagick/utils.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
+package imagick
+
+import (
+	"github.com/journeymidnight/pipa/helper"
+	"github.com/journeymidnight/pipa/library"
+	"gopkg.in/gographics/imagick.v3/imagick"
+)
+
+func adjustCropTask(plan ResizePlan, width, height uint) {
+	//单宽高缩放
```
Comment (fireworkmarks) [inline imagick/utils.go]: same

File: imagick/watermark.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+	fill     bool
+}
+
+func newWatermark() Watermark {
+	return Watermark{library.XMargin, library.YMargin, GRAVITY, library.Transparency, nil, new(Text)}
+}
+
+func (img *ImageWand) watermark(o Watermark) (err error) {
+	if o.Picture != nil {
+		err = img.MagickWand.CompositeImage(o.Picture, o.Picture.GetImageCompose(), true, o.XMargin, o.YMargin)
```
Comment (fireworkmarks) [inline imagick/watermark.go:47]: You only need to pass the o parameter, and the abbreviation should not be o，m is better

File: library/library.go [hunk 10]
```diff
@@ -2,6 +2,45 @@ package library
...
+	Force      = false
+	Crop       = false
+	Pad        = false
+	Limit      = true
+	//Watermark default param
+	XMargin      = 10
+	YMargin      = 10
+	Transparency = 100
+	FrontSize    = 40.0
+	Background   = "#FFFFFF"
```
Comment (fireworkmarks) [inline library/library.go]: What is this value? Is it the default value?

File: pipa/util.go [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+	pos := strings.Index(urlFragments[0], OssProcess)
+	if pos == -1 {
+		return "", operations, ErrNotFoundOssProcess
+	}
+
+	path, err := url.Parse(urlFragments[0])
+	if err != nil {
+		return "", operations, err
+	}
+	host := "http://" + path.Hostname()
```
Comment (fireworkmarks) [inline pipa/util.go]: const

File: go.mod [hunk 12]
```diff
@@ -0,0 +1,9 @@
+module github.com/journeymidnight/pipa
+
+require (
+	github.com/BurntSushi/toml v0.3.1
+	github.com/garyburd/redigo v1.6.0
+	gopkg.in/gographics/imagick.v3 v3.2.0
+)
+
+go 1.12
```
Comment (fireworkmarks) [inline go.mod]: change go 1.13

File: imagick/process.go [hunk 13]
```diff
@@ -29,3 +86,233 @@ func NewImageWand() (lib Library) {
 	}
 	return &img
 }
+
+func (img *ImageWand) ResizeImageProcess(data []byte, plan ResizePlan) error {
+	helper.Log.Info("start resize image, plan: ", )
+	err := img.MagickWand.ReadImageBlob(data)
+	if err != nil {
+		helper.Log.Error("read data failed")
```
Comment (fireworkmarks) [inline imagick/process.go]: If you want to log err, put err stuff behind

File: error/error_codes.go [hunk 14]
```diff
@@ -21,6 +29,38 @@ var ErrorCodeResponse = map[PipaError]ErrorStruct{
...
+		ErrorCode:    402,
+		ErrorMessage: "Can not parameter x-oss-process.",
+	},
+	ErrInvalidParameter: {
+		ErrorCode:    403,
+		ErrorMessage: "Invalid parameter.",
+	},
+	ErrInvalidParameterFormat: {
+		ErrorCode:    405,
+		ErrorMessage: "Invalid parameter.",
```
Comment (fireworkmarks) [inline error/error_codes.go]: Better not have the same information

File: handle/worker.go [hunk 15]
```diff
@@ -107,22 +111,60 @@ func slave(slave_num int) {
...
+	mimeType := resp.Header.Get("Content-Type")
+
+	if strings.Contains(mimeType, "image") == false {
+		if ok, _ := regexp.MatchString("(jpeg|jpg|png|gif|bmp|webp|tiff)", downloadUrl); ok == false {
+			helper.Log.Info(fmt.Sprintf("MIME TYPE is %s not an image\n", mimeType))
+			return nil, StatusUnsupportedMediaType //415
+		}
+	}
+	contentLength := resp.Header.Get("Content-Length")
+	if len, _ := strconv.Atoi(contentLength); len > (20 << 20) {
```
Comment (fireworkmarks) [inline handle/worker.go]: Involving numerical values, especially the left shift number is defined with const

File: error/error_codes.go [hunk 16]
```diff
@@ -21,14 +48,122 @@ var ErrorCodeResponse = map[PipaError]ErrorStruct{
...
+		ErrorMessage: "Watermark picture download failed!.",
+	},
 }
 
-func (e PipaError) ErrorCode() int {
+func (e PipaError) ErrorCode() (int, string) {
 	err, ok := ErrorCodeResponse[e]
 	if !ok {
-		return 400
+		return 400, "No error has found"
```
Comment (cannium) [inline error/error_codes.go:164]: 400 and "no error has found"...

File: imagick/watermark.go [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+	fill     bool
+}
+
+func newWatermark() Watermark {
+	return Watermark{XMargin, YMargin, DdfaultGravity, Transparency, nil, new(Text)}
+}
+
+func (img *ImageWand) watermark(o Watermark) (err error) {
+	if o.Picture != nil {
+		err = o.Picture.SetImageAlphaChannel(AlphaOpaque)
```
Comment (cannium) [inline imagick/watermark.go:37]: What does this command do?

File: handler/worker.go [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+const TaskQueue = "taskQueue"
+const PictureMAxSize = 20 << 20
```
Comment (cannium) [inline handler/worker.go:20]: typo

File: handler/worker.go [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+		if ok, _ := regexp.MatchString("(jpeg|jpg|png|gif|bmp|webp|tiff)", downloadUrl); ok == false {
+			helper.Log.Info(fmt.Sprintf("MIME TYPE is %s not an image\n", mimeType))
+			return nil, StatusUnsupportedMediaType //415
+		}
+	}
+	contentLength := resp.Header.Get("Content-Length")
+	if len, _ := strconv.Atoi(contentLength); len > PictureMAxSize {
+		return nil, StatusRequestEntityTooLarge
+	}
+
```
Comment (cannium) [inline handler/worker.go:161]: Put a `LimitedReader` before `resp.Body`, like what Yig does.

File: handler/worker.go [hunk 20]
```diff
@@ -0,0 +1,242 @@
...
+			}
+			wg.Done()
+		}
+	}
+}
+
+func downloadImage(downloadUrl string) ([]byte, error) {
+	helper.Log.Info(fmt.Sprintf("Start to download %s\n", downloadUrl))
+
+	httpClient := &http.Client{Timeout: time.Second * 5}
```
Comment (cannium) [inline handler/worker.go:134]: 5 sec might be too small, it includes transport time of the request body, which could be 20M.

File: handler/worker.go [hunk 21]
```diff
@@ -0,0 +1,242 @@
...
+	}
+	go listenFinishedTask(ReturnQ)
+
+	for {
+		t, err := receiveImageTask()
+		if err != nil {
+			continue
+		}
+		helper.Log.Info("receive image task:", t)
+		TaskQ <- t
```
Comment (cannium) [inline handler/worker.go:70]: This line would panic if `TaskQ` is closed.

File: handler/worker.go [hunk 22]
```diff
@@ -0,0 +1,242 @@
...
+	}, nil
+}
+
+func listenFinishedTask(resultQ chan FinishedTask) {
+	c := redis.Pool.Get()
+	defer c.Close()
+	for r := range resultQ {
+		//put data back to redis
+		if r.code == 200 {
+			_, err := c.Do("MULTI")
```
Comment (cannium) [inline handler/worker.go:189]: Do we really need a transaction here? If so, execute `DISCARD` on error cases.

File: imagick/resize.go [hunk 23]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+func newResize() *Resize {
+	return &Resize{0, 0, Zoom, Force, Crop, Pad, Limit, Background}
+}
+
+func imageCalculations(o *Resize, inWidth, inHeight int) float64 {
+	factor := 1.0
+	wFactor := float64(o.Width) / float64(inWidth)
+	hFactor := float64(o.Height) / float64(inHeight)
```
Comment (cannium) [inline imagick/resize.go:88]: `inWidth` and `inHeight` could be zero?

File: imagick/utils.go [hunk 24]
```diff
@@ -0,0 +1,164 @@
+package imagick
+
+import (
+	"github.com/journeymidnight/pipa/helper"
+	"gopkg.in/gographics/imagick.v3/imagick"
+)
+
+const (
+	//Resize default param
+	Zoom       = 0.0
```
Comment (cannium) [inline imagick/utils.go:10]: Default to 1.0?

File: handler/util.go [hunk 25]
```diff
@@ -0,0 +1,162 @@
...
+func getKeyAndValue(paramKeys []string) (captures map[string]string, err error) {
+	captures = make(map[string]string)
+	for _, param := range paramKeys {
+		if param == "" {
+			return captures, ErrInvalidParametersHaveSpaces
+		}
+		keys := strings.Split(param, "_")
+		//if len(keys) > 2 {
+		//	return captures, ErrInvalidParameterFormat
+		//}
```
Comment (fireworkmarks) [inline handler/util.go:111]: Why not judge

File: imagick/resize.go [hunk 26]
```diff
@@ -0,0 +1,139 @@
...
+
+	switch {
+	case o.Width > 0 && o.Height > 0:
```
Comment (fireworkmarks) [inline imagick/resize.go:91]: Suggest final judgment

**Pull request merged**